File: bindings/node/__tests__/backtest.test.js
const { test } = require('node:test')
const assert = require('node:assert')
const { run, version } = require('../index.js')

const PRICE_SPEC = JSON.stringify({
  symbol: 'x',
  timeframe: '1h',
  indicators: {},
  entry: { gt: [{ price: 'close' }, 100] },
  exit: { lt: [{ price: 'close' }, 100] },
  sizing: { type: 'fixed_qty', qty: 1 },
})

test('version is a non-empty string', () => {
  assert.ok(typeof version() === 'string' && version().length > 0)
})

test('hand-computed round trip matches the Rust engine', () => {
  const open = [100, 102, 104, 98]
  const high = [101, 103, 104, 98]
  const low = [100, 102, 99, 97]
  const close = [101, 103, 99, 97]
  const volume = [0, 0, 0, 0]
  const time = [0, 1, 2, 3]
  const r = JSON.parse(run(open, high, low, close, volume, time, PRICE_SPEC, 1000))
  assert.strictEqual(r.metrics.num_trades, 1)
  assert.ok(Math.abs(r.trades[0].entry_price - 102) < 1e-9)
  assert.ok(Math.abs(r.trades[0].exit_price - 98) < 1e-9)
  assert.ok(Math.abs(r.trades[0].pnl - -4) < 1e-9)
  assert.ok(Math.abs(r.equity[r.equity.length - 1].equity - 996) < 1e-9)
})

test('ema crossover runs', () => {
  const spec = JSON.stringify({
    symbol: 'x',
    timeframe: '1h',
    indicators: { fast: { type: 'Ema', params: [5] }, slow: { type: 'Ema', params: [15] } },
    entry: { cross_above: ['fast', 'slow'] },
    exit: { cross_below: ['fast', 'slow'] },
    sizing: { type: 'fixed_fraction', fraction: 0.5 },
  })
  const n = 60
  const close = Array.from({ length: n }, (_, i) => 100 + 10 * Math.sin(i / 3))
  const high = close.map((c) => c + 1)
  const low = close.map((c) => c - 1)
  const volume = new Array(n).fill(0)
  const time = Array.from({ length: n }, (_, i) => i)
  const r = JSON.parse(run(close, high, low, close, volume, time, spec))
  assert.strictEqual(r.equity.length, n)
  assert.strictEqual(r.schema_version, 1)
})

test('invalid spec throws', () => {
  assert.throws(() => run([1], [1], [1], [1], [0], [0], JSON.stringify({ not: 'valid' })))
})

File: bindings/go/backtest_test.go
package wickrabacktest

import (
	"encoding/json"
	"math"
	"testing"
)

const priceSpec = `{"symbol":"x","timeframe":"1h","indicators":{},` +
	`"entry":{"gt":[{"price":"close"},100]},` +
	`"exit":{"lt":[{"price":"close"},100]},` +
	`"sizing":{"type":"fixed_qty","qty":1}}`

func TestVersionIsNonEmpty(t *testing.T) {
	if Version() == "" {
		t.Fatal("version is empty")
	}
}

func TestHandComputedRoundTripMatchesEngine(t *testing.T) {
	open := []float64{100, 102, 104, 98}
	high := []float64{101, 103, 104, 98}
	low := []float64{100, 102, 99, 97}
	close := []float64{101, 103, 99, 97}
	time := []int64{0, 1, 2, 3}

	out, err := Run(open, high, low, close, nil, time, priceSpec, 1000.0)
	if err != nil {
		t.Fatalf("Run failed: %v", err)
	}

	var report struct {
		Metrics struct {
			NumTrades int `json:"num_trades"`
		} `json:"metrics"`
		Trades []struct {
			EntryPrice float64 `json:"entry_price"`
			ExitPrice  float64 `json:"exit_price"`
			Pnl        float64 `json:"pnl"`
		} `json:"trades"`
		Equity []struct {
			Equity float64 `json:"equity"`
		} `json:"equity"`
	}
	if err := json.Unmarshal([]byte(out), &report); err != nil {
		t.Fatalf("bad report json: %v", err)
	}

	if report.Metrics.NumTrades != 1 {
		t.Fatalf("num_trades = %d, want 1", report.Metrics.NumTrades)
	}
	tr := report.Trades[0]
	if math.Abs(tr.EntryPrice-102.0) > 1e-9 {
		t.Errorf("entry_price = %v, want 102", tr.EntryPrice)
	}
	if math.Abs(tr.ExitPrice-98.0) > 1e-9 {
		t.Errorf("exit_price = %v, want 98", tr.ExitPrice)
	}
	if math.Abs(tr.Pnl-(-4.0)) > 1e-9 {
		t.Errorf("pnl = %v, want -4", tr.Pnl)
	}
	last := report.Equity[len(report.Equity)-1].Equity
	if math.Abs(last-996.0) > 1e-9 {
		t.Errorf("final equity = %v, want 996", last)
	}
}

func TestInvalidSpecReturnsError(t *testing.T) {
	one := []float64{1.0}
	if _, err := RunSimple(one, one, one, one, `{"bad":true}`, 10_000.0); err == nil {
		t.Fatal("expected error for invalid spec")
	}
}
